bus-mapping/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+//! Crate docs
+
+#![cfg_attr(docsrs, feature(doc_cfg))]
+// Temporary until we have more of the crate implemented.
+#![allow(dead_code)]
+// Catch documentation errors caused by code changes.
+#![deny(broken_intra_doc_links)]
+#![deny(missing_debug_implementations)]
+#![deny(missing_docs)]
+#![deny(unsafe_code)]
+
+use std::{
+    collections::BTreeMap,
+    iter::Iterator,
+    ops::{Index, IndexMut},
+};
+
+use itertools::Itertools;
+use pasta_curves::arithmetic::FieldExt;
+
+// -------- EVM Circuit
+// gc as index of each row corresponds to the state after the opcode is performed
+// 1 gc has 1 entry
+//`StackElem{key 	val 	stack_op(READ/WRITE) 	gc   OPCODE    CallID(later_future)}`
+
+//------ State Circuit
+// Group target (stack or memory)
+// Group by key (location in memory/storage)
+// Sorty by gc
+//`MemoryElem{target 	gc 	val1 	val2 	val3}`
+
+#[derive(Debug, Clone)]
+struct BlockConstants<F: FieldExt> {
+    hash: [u8; 256], // Until we know how to deal with it
+    coinbase: F,
+    timestamp: F,
+    number: F,
+    difficulty: F,
+    gaslimit: F,
+    chain_id: F,
+}
+
+#[derive(Debug, Clone)]
+enum RW {
+    READ,
+    WRITE,
+}
+
+/// Doc
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Target {
+    /// Doc
+    Memory,
+    /// Doc
+    Stack,
+    /// Doc
+    Storage,
+}
+
+/// Doc
+#[derive(Debug, Clone)]
+pub struct Operation<F: FieldExt> {
+    mem_op: RW,
+    target: Target,
+    key: usize,
```

Comment:
With more understanding about evm, I found evm uses `keccack(...)` as storage location a lot. So I am afraid that `usize` could not fit a 32-byte word for target storage.

But it fits stack and memory usage for now, maybe we can leave it for future PR?

Reply:
you are absolutely right.. It will need to be `F` anyway at the end. I'm sure about it.. The only problem is that as you say It's sad that just for the storage case we need to store a Bn Scalar and we cannot simply store the usize can convert to Bn Point when returning the iterator...

I'll address it on this PR I think since this could lead to issues IMO. Or no? WDYT? We should try to be consistent even if the storage part hasn't been developed yet.

Reply:
I think we should have u256 instead of `F` for storage location because the original value of location is important and should be in evm circuit byte to byte, although we might use the compressed form in state circuit.

Do you think introducing a new struct for each enum variant of `Target` acceptable? For example, we will have:
```rust
struct Stack(usize);
struct Memory(usize);
struct Storage(SomeU256Type);

enum Target {
  Stack(Stack),
  Memory(Memory),
  Storage(Storage),
}

impl Target {
  fn stack(&self) -> Option<Stack> { ... }
  fn memory(&self) -> Option<Memory> { ... }
  fn storage(&self) -> Option<Storage> { ... }
}
```

Reply:
I thought the storage words were always a BN256 element. If it's not the case, Idk wether is better to store as U256 or directly as [u8;32] since the second one already implements `Copy` and we can do much more stuff over it rather than over a custom type. 

It would make sense to have `SomeU256Type` if we needed to operate with it, but AFAIK that's not the case, and we just need it to hash it or add it into lookup tables, which should not be an issue if we have it as a byte array.

WDYT @han0110 ?